expo/notifications.go

```diff
@@ -0,0 +1,198 @@
+}
+
+func (r *PushResponse) isSuccess() bool {
+	return r.Status == SuccessStatus
+}
+
+func (r *PushResponse) ValidateResponse() error {
+	if r.isSuccess() {
+		return nil
+	}
+	err := &PushResponseError{
+		Response: r,
+	}
+	// Handle specific errors if we have information
+	if r.Details != nil {
+		e := r.Details["error"]
+		if e == ErrorDeviceNotRegistered {
+			return &DeviceNotRegisteredError{
+				PushResponseError: *err,
+			}
+		} else if e == ErrorMessageTooBig {
+			return &MessageTooBigError{
+				PushResponseError: *err,
+			}
+		} else if e == ErrorMessageRateExceeded {
+			return &MessageRateExceededError{
+				PushResponseError: *err,
+			}
+		}
+	}
+	return err
+}
+
+// SuccessStatus is the status returned from Expo on a success
+const SuccessStatus = "ok"
+
+// ErrorDeviceNotRegistered indicates the token is invalid
+const ErrorDeviceNotRegistered = "DeviceNotRegistered"
+
+// ErrorMessageTooBig indicates the message went over payload size of 4096 bytes
+const ErrorMessageTooBig = "MessageTooBig"
+
+// ErrorMessageRateExceeded indicates messages have been sent too frequently
+const ErrorMessageRateExceeded = "MessageRateExceeded"
+
+type PushResponseError struct {
+	Response *PushResponse
+}
+
+func (e *PushResponseError) Error() string {
+	if e.Response != nil {
+		return e.Response.Message
+	}
+	return "Unknown push response error"
+}
+
+// DeviceNotRegisteredError is raised when the push token is invalid
+// To handle this error, you should stop sending messages to this token.
+type DeviceNotRegisteredError struct {
+	PushResponseError
+}
+
+// MessageTooBigError is raised when the notification was too large.
+// On Android and iOS, the total payload must be at most 4096 bytes.
+type MessageTooBigError struct {
+	PushResponseError
+}
+
+// MessageRateExceededError is raised when you are sending messages too frequently to a device
+// You should implement exponential backoff and slowly retry sending messages.
+type MessageRateExceededError struct {
+	PushResponseError
+}
+
+type PushServerError struct {
+	Message      string
+	Response     *http.Response
+	ResponseData *Response
+	Errors       []map[string]string
+}
+
+// NewPushServerError creates a new PushServerError object
+func NewPushServerError(message string, response *http.Response,
+	responseData *Response,
+	errors []map[string]string) *PushServerError {
+	return &PushServerError{
+		Message:      message,
+		Response:     response,
+		ResponseData: responseData,
+		Errors:       errors,
+	}
+}
+
+func (e *PushServerError) Error() string {
+	return e.Message
+}
+
+func checkStatus(resp *http.Response) error {
+	if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
+		return nil
+	}
+	return fmt.Errorf("invalid response (%d %s)", resp.StatusCode, resp.Status)
+}
+
+func PublishMessages(messages []PushMessage) ([]PushResponse, error) {
+	path := "https://exp.host/--/api/v2/push/send"
+	token := os.Getenv("EXPO_PUSH_ACCESS_TOKEN")
```

Comment:
I'd recommend making callers provide this as an argument. Ideally only package main calls os.Getenv.

Reply:
I'm using the official expo package now for this functionality.